addons/product/product.py

```diff
@@ -935,6 +935,45 @@ def _get_price_extra(self, cr, uid, ids, name, args, context=None):
             result[product.id] = price_extra
         return result
 
+    def _set_suppliers(self, cr, uid, id, field_name, field_value, args, context=None):
+        if not field_value:
+            return
+        product_tmpl_id = self.browse(cr, uid, id, context=context).product_tmpl_id.id
+        supplierinfo = self.pool['product.supplierinfo']
+        for value in field_value:
+            if value[0] == 0:
+                value[2].update(product_tmpl_id=product_tmpl_id, product_id=id)
+                supplierinfo.create(cr, uid, value[2], context=context)
+            elif value[0] == 1:
+                if not supplierinfo.browse(cr, uid, value[1], context=context).product_id:
+                    raise UserError(_("You are not allowed to modify the attributes of product template's suppliers."))
```

Comment:
actually, the message is referring the product.template record rather than product.product, hence 'product_tmpl.name' is enough.